src/mscorlib/shared/System/Text/StringBuilder.cs

```diff
@@ -790,6 +790,109 @@ public StringBuilder Append(string value, int startIndex, int count)
             }
         }
 
+        public StringBuilder Append(StringBuilder value)
+        {
+            if (value != null && value.Length != 0)
+            {
+                if (value == this)
+                    Append(value.ToString());
+                else
+                    AppendCore(value, 0, value.Length);
+            }
+            return this;
+        }
+
+        public StringBuilder Append(StringBuilder value, int startIndex, int count)
+        {
+            if (startIndex < 0)
+            {
+                ThrowHelper.ThrowStartIndexArgumentOutOfRange_ArgumentOutOfRange_Index();
+            }
+
+            if (count < 0)
+            {
+                ThrowHelper.ThrowArgumentOutOfRangeException(ExceptionArgument.count, ExceptionResource.ArgumentOutOfRange_NeedNonNegNum);
+            }
+
+            if (value == null)
+            {
+                if (startIndex == 0 && count == 0)
+                {
+                    return this;
+                }
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.value);
+            }
+
+            if (count > value.Length - startIndex)
+            {
+                ThrowHelper.ThrowStartIndexArgumentOutOfRange_ArgumentOutOfRange_Index();
+            }
+
+            if (count > 0)
+            {
+                if (value == this)
+                    Append(value.ToString(startIndex, count));             
+                else
+                    AppendCore(value, startIndex, count);
```

Comment:
Nit: It should be slightly better to do `return AppendCore` here and other similar places in this file because of it lets the JIT to employ [tailcall](https://en.wikipedia.org/wiki/Tail_call) optimization and not worry about keeping `this` pointer alive.